### DISCUSSION TITLE
guide-R/analyses/regression-logistique-binaire

### DISCUSSION THREAD

**Discussion by giscus** [larmarange/guide-R#23] [Forum]
# guide-R/analyses/regression-logistique-binaire

https://larmarange.github.io/guide-R/analyses/regression-logistique-binaire.html

**Comment by babelmansour2** [comment]
Bonjour Joseph, je conduis une régression ordinale avec trois groupes. J'obtiens bien sûr les OR et surtout (grâce à vous) les probabilités pour chaque individu d'appartenir à tel ou tel groupe avec le package marginaleffects. Mais je voudrais les "chances" en pourcentage d'appartenir à tel ou tel groupe selon, par exemple, que l'on est un homme ou une femme, soit l'équivalent de ce que fait logit_inverse pour les régressions logistiques binaires. Merci à vous !

**Reply by larmarange** [reply]
Bonjour,

je ne suis pas bien sur de comprendre la question. N'est-ce pas ce que vous obtenez avec les prédictions marginales ?

Eventuellement, pourriez-vous reformuler votre question avec un exemple concret, en prenant pour appui le modèle présenté ici : https://larmarange.github.io/guide-R/analyses_avancees/regression-logistique-ordinale.html

Cdlt

**Comment by babelmansour2** [comment]
Bonjour Joseph, merci de votre réponse. J'ai lu bien sûr votre chapitre Guide R sur le sujet. Si je reprends votre exemple, je voudrais calculer les chances en pourcentage des hommes et des femmes d'appartenir à telle ou telle catégorie de la variable indépendante (dans votre exemple la satisfaction au travail), comme vous le faites avec la fonction logit_inverse pour la régression logistique. Autrement dit, convertir les odd ratios en pourcentage. Je ne sais pas si je suis bien clair. Bien à vous

**Reply by babelmansour2** [reply]
Telle ou telle catégorie de la variable dépendante, excusez-moi !

**Reply by larmarange** [reply]
Reprenons le modèle rego2 présenté https://larmarange.github.io/guide-R/analyses_avancees/regression-logistique-ordinale.html

Voici les données du modèle brut:

```
> rego2
Call:
MASS::polr(formula = trav.satisf ~ etudes + trav.imp, data = d)

Coefficients:
               etudesSecondaire etudesTechnique / Professionnel                 etudesSupérieur             etudesNon documenté 
                     0.06066981                      0.01390364                      0.60999364                     -0.41614548 
                  trav.impAussi                   trav.impMoins                     trav.impPeu 
                     0.33213429                      0.03320042                     -1.63551894 

Intercepts:
Insatisfaction|Equilibre   Equilibre|Satisfaction 
              -1.9463675                0.4129235 

Residual Deviance: 1958.973 
AIC: 1976.973 
(952 observations effacées parce que manquantes)
```

Nous voyons qu'il y a deux intercepts.

Regardons ce qu'il se passe à la référence du niveau d'étude et à la référence de l'importance donnée au travail.

L'intercept `Insatisfaction|Equilibre` correspond au fait d'être insatisfait (ou moins mais c'est réthorique) vs à l'équilibre ou plus (ie équilibre ou satisfait). Donc pour la probabilité d'être Insatisfait, `logit_inverse(-1.9463675)` ce qui donne 12,5%.

L'intercept `Equilibre|Satisfiaction` correspond au fait d'être à l'Equilibre (ou moins soit Insatisfait) vs Satisfait ou plus. Donc la probabilité d'être à l'équilibre ou insatisfait est `logit_inverse(1958.973)` soit 60,2%. On peut en déduire que la probabilité d'être Satisfait est donc de 39,8%. 

Et celle d'être à l'équilibre est `logit_inverse(0.4129235) - logit_inverse(-1.9463675)` soit 47.7%.

Cela est valable pour la situation à la référence (i.e. Niveau primaire et Importance Le plus). Si ensuite on veut calculer la situation pour d'autres combinaisons de modalités, il faut ajouter les coefficients des termes (qui sont identiques quel que soit le niveau).

Une manière plus simple, est d'avoir recours à `broom.helpers::tidy_ggpredict()`

```
> rego2 |> broom.helpers::tidy_ggpredict()
# A tibble: 27 × 7
   y.level        variable term                      estimate std.error conf.low conf.high
   <chr>          <chr>    <chr>                        <dbl>     <dbl>    <dbl>     <dbl>
 1 Insatisfaction etudes   Primaire                    0.125      0       0.125      0.125
 2 Equilibre      etudes   Primaire                    0.477      0       0.477      0.477
 3 Satisfaction   etudes   Primaire                    0.398      0       0.398      0.398
 4 Insatisfaction etudes   Secondaire                  0.118      0.232   0.0785     0.175
 5 Equilibre      etudes   Secondaire                  0.469      0.232   0.359      0.582
 6 Satisfaction   etudes   Secondaire                  0.413      0.232   0.308      0.526
 7 Insatisfaction etudes   Technique / Professionnel   0.123      0.211   0.0851     0.176
 8 Equilibre      etudes   Technique / Professionnel   0.475      0.211   0.374      0.578
 9 Satisfaction   etudes   Technique / Professionnel   0.402      0.211   0.307      0.504
10 Insatisfaction etudes   Supérieur                   0.0720     0.220   0.0480     0.107
# ℹ 17 more rows
# ℹ Use `print(n = ...)` to see more rows
```

Ici, les prédictions pour les modalités d'une variable se font à la référence des autres variables.

C'est une forme de prédiction marginale.

**Reply by babelmansour2** [reply]
Merci beaucoup ! J'aurais aimé vous avoir pour professeur de statistiques ;o) Apparemment la fonction logit_inverse n'existe plus dans R, du moins je ne l'ai pas trouvée même en chargeant tous les packages utilisés, ni sur Internet. Mais on peut la recréer pour ceux que cela intéresserait : 
logit_inverse <- function(logit_values) {
  odds <- exp(logit_values)
  probabilities <- odds / (1 + odds)
  return(probabilities)
}

**Reply by larmarange** [reply]
Pardon, la fonction logit_inverse() est présentée ici https://larmarange.github.io/guide-R/analyses/regression-logistique-binaire.html#sec-interpreter-coefficients-regression-logistique

On peut l'obtenir facilement avec 

`logit_inverse <- binomial("logit") |> purrr::pluck("linkinv")`

**Reply by babelmansour2** [reply]
Merci à vous !

**Comment by WilliamZ98** [comment]
Bonjour monsieur, merci pour ce cours. J’ai des questions :

-	En ce qui concerne le recode de l’âge en classe, pourquoi vous n’avez pas opté pour des classes suivants les quartiles ?
-	Qu’est-ce qu’il faut comprendre par modalité sous-représentée ? est-ce au critère de 5% que l’on tranche ?
-	Entre la statistique bivariée et la régression univariée, qu’est-ce qu’il faut utiliser et/ou présenter pour la sélection des variables à inclure dans le modèle multivarié ?
-	Ici qu’est-ce qui justifie le choix du modèle avec la fonction de lien logit ? Et, quelle syntaxe en R permettrait de réaliser une comparaison de modèles (l’un logit et l’autre probit) sur la base des critères d’informations, de la vraisemblance, etc. C’est dire avoir ces résultats dans un seul et même tableau.
-	La mise en garde sur l’interprétation de l’odds ratio, je ne l’ai pas vraiment compris. Puis-je avoir des ressources sur l’interprétation des odds ratio et la différence avec la prévalence ratio ?

Respectueusement !

**Reply by larmarange** [reply]
> En ce qui concerne le recode de l’âge en classe, pourquoi vous n’avez pas opté pour des classes suivants les quartiles ?

IL n'y a pas de règle absolue. Ce qui compte c'est que cela fasse sens du point des questions scientifiques posées et de l'objet de recherche. Ici, j'ai préféré des classes d'âges qui correspondent à des situations sociologiques : 18-24 ans est la tranche d'âges usuellement utilisée pour définir les jeunes, et 65 ans et plus permet de capter la population retraitée.

> Qu’est-ce qu’il faut comprendre par modalité sous-représentée ? est-ce au critère de 5% que l’on tranche ?

Encore une fois, il n'y a pas de règle absolue et les arbitrages sont fonction des questions que l'on se pose. Ceci dit, si le nombre d'observations dans la modalité est faible (<50), on manquera de puissance statistique pour identifier un éventuel effet.

De manière générale, il est souvent préférable de simplifier les variables avant régression, quitte à rediviser une variable d'intérêt particulière dans un second temps si l'on a des hypothèses spécifiques.

> Entre la statistique bivariée et la régression univariée, qu’est-ce qu’il faut utiliser et/ou présenter pour la sélection des variables à inclure dans le modèle multivarié ?

En premier lieu, la liste des facteurs considérés doit se baser sur une revue de la littérature ou des hypothèses spécifiques sur le phénomène étudié. On ne "balance" pas toutes les variables disponibles dans un modèle.

Ensuite, il est important de rendre compte de toutes les variables considérées pour l'analyse au moment de l'analyse bivariée, même si juste une partie est finalement retenu pour l'analyse multivariée. Il est tout aussi important de savoir qu'une variable n'a pas été retenue que de savoir qu'elle a été incluse dans le modèle.

Si l'on a recours à une approche pour sélectionner seulement certaines variables dans le modèle finale (avec un step() ou par exemple en ne retenant que celle ayant une p-valeur < 20% en bivariée), celle-ci doit être explicitée. Cela fait partie de la méthodologie à rapporter dans un article. Tous les choix doivent être explicités.

> Ici qu’est-ce qui justifie le choix du modèle avec la fonction de lien logit ? Et, quelle syntaxe en R permettrait de réaliser une comparaison de modèles (l’un logit et l’autre probit) sur la base des critères d’informations, de la vraisemblance, etc. C’est dire avoir ces résultats dans un seul et même tableau.

L'intérêt du modèle logit est que l'exponentielle des coefficients peut s'interpréter comme des odds ratio. Ce qui n'est pas la cas d'un modèle probit. 

Un modèle probit s'obtient avec `family = binomial("probit")`. À l'usage, pour ma part, je n'ai jamais eu besoin de modèles probit et j'ai toujours privilégié les modèles logit.

> La mise en garde sur l’interprétation de l’odds ratio, je ne l’ai pas vraiment compris. Puis-je avoir des ressources sur l’interprétation des odds ratio et la différence avec la prévalence ratio ?

Supposons que vous ayez deux groupes (A et B) où les proportions sont respectivement de 40% et 20%. Le prévalence ratio est de 0.4/0.2 = 2 tandis que l'odds ratio est de (0.4/0.6)/(0.2/0.8) = 2.667. Les deux valeurs ne sont pas identiques et il serait faux de dire que le groupe A a 2,6 plus de chance de vivre l'événement que le groupe B. Il a 2 fois plus de chance. Un OR ne s'interprète donc pas exactement comme un PR (sauf si le phénomène étudié à une faible proportion, <5%, auquel cas les deux valeurs sont très proches).

Si l'on a besoin de PR, on aura recours à un modèle de comptage sur variable binaire, cf. https://larmarange.github.io/guide-R/analyses_avancees/modeles-comptage.html#mod%C3%A8les-de-comptage-avec-une-variable-binaire

Il y a de la littérature sur ce sujet, par exemple https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5135596/ ou https://www.frontiersin.org/articles/10.3389/fvets.2017.00193/full ou https://bmcmedresmethodol.biomedcentral.com/articles/10.1186/1471-2288-3-21 pour ne prendre que les premiers résultats d'une recherche sur le web

**Reply by WilliamZ98** [reply]
Merci pour votre réponse !

**Comment by Bolenli** [comment]
Bonjour Joseph , et merci pour ce cours sur la régression logistique binaire qui personnellement m'a beaucoup aidé dans mon travail sur R. Cependant j'ai une question sur le calcul des probabilités. En fonction des données des OR dur le tableau ci dessous sur la pratique du sport, comment m'y prendre pour avoir la probabilité pour une femme  âgée de 25-44 ans, de niveau secondaire, pratiquant occasionnellement? ainsi que la probabilité pour un homme ayant les mêmes caractéristiques que  celles de la femme citées? Je vous remercie

**Reply by larmarange** [reply]
L'approche de base consiste à prendre les coefficients bruts (pas les OR). Vous partez de l'intercept et vous lui additionnez tous les coefficients qui s'appliquent puis vous appliquer la fonction `logit_inverse()` définie dans le chapitre.

De manière plus générique, vous faites un tableau de données à une ligne avec les modalités voulues (voir exemple avec `individus3` dans le chapitre) et vous utilisez la fonction `predict()` avec l'option `type = "response"`

**Reply by Bolenli** [reply]
Bonjour Joseph. Merci c'est noté

**Comment by Ineschalet** [comment]
Bonjour,
Merci beaucoup ce cours. J'aimerai savoir s'il était possible d'avoir l'odd ration pour une unité différente de 1 pour les variables quantitatives. Par exemple pour les heures.tv ici cela va indiquer un odd ration pour une augmentation de 1 unité. Comment faire pour que ça soit 2 ou 3 ?
Merci à vous !

**Reply by larmarange** [reply]
Vous pouvez le calculer manuellement : si votre OR est de 1,4,  l'OR pour 3 heures est 1,4^3=1,4x1,4x1,4=2,744

**Reply by larmarange** [reply]
Ou alors, en amont, vous créez une variable `htv3 = heures.tv / 3` et c'est cette nouvelle variable que vous utilisez dans le modèle.

Ou encore en faisant le calcul directement dans l'équation du modèle : `mod <- glm(sport ~ I(heures.tv / 3) + sexe, family = binomial, data = hdv2003)
`

**Comment by Melodie-Clvr** [comment]
Bonjour, 
Merci pour ce cours. Reprenant votre exemple, est-il possible d'obtenir une p_value lors de la comparaison des odds ratio : d'un homme de 18-24 ans dans le supérieur et pratiquant occasionnel vs une femme de 65 ans et plus non documenté sans croyance ni appartenance ?
Merci

**Reply by larmarange** [reply]
Il me semble que votre question est mal formulée. Vous ne souhaitez pas ici tester si un Odds ratio est différent de 1, mais comparer deux profils (ici vous faites 4 variables et non pas une seule).

Pour cela, il vous jeter un email au package `emmeans voir https://aosmith.rbind.io/2019/04/15/custom-contrasts-emmeans/`

**Reply by Melodie-Clvr** [reply]
D'accord, merci de votre réponse !
